Optimize DefaultThreadContextMap.getCopy() performance

anuragdy opened the pull request:
This PR optimizes the `DefaultThreadContextMap.getCopy()` method to address performance issues related to megamorphic calls in the HashMap constructor, providing **30-50% performance improvements** for non-empty maps.

The original implementation used `new HashMap<>(map)` which suffers from megamorphic call performance issues documented in:
- **GitHub Issue**: #3935  
- **JDK Ticket**: [JDK-8368292](https://bugs.openjdk.org/browse/JDK-8368292)

The HashMap constructor with a Map parameter requires (3 + 4n) virtual method calls that become megamorphic when used with different map types, leading to 24-136% performance degradation compared to manual iteration.

Closes #3935.

anuragdy committed:
Optimize DefaultThreadContextMap.getCopy() performance

vy commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/DefaultThreadContextMap.java:
```diff
@@ -147,13 +147,53 @@ public <V> V getValue(final String key) {
+        // Manual iteration avoids megamorphic virtual calls that prevent JIT optimization.
+        // The HashMap(Map) constructor requires (3 + 4n) virtual method calls that become
+        // megamorphic when used with different map types, leading to 24-136% performance
+        // degradation. Manual iteration creates monomorphic call sites that JIT can optimize.
```
```suggestion
        // degradation. Manual iteration creates monomorphic call sites that JIT can optimize.
        // See https://bugs.openjdk.org/browse/JDK-8368292
```

vy commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/DefaultThreadContextMap.java:
```diff
@@ -147,13 +147,53 @@ public <V> V getValue(final String key) {
     }
 
+    /**
+     * Returns a mutable copy of the current thread context map.
+     * <p>
+     * This method has been optimized to avoid performance issues with the HashMap(Map) constructor
+     * that suffers from megamorphic call overhead (see JDK-8368292 and GitHub issue #3935).
+     * The optimization provides 30-50% performance improvement for non-empty maps by using
+     * manual iteration instead of the HashMap constructor.
+     * </p>
+     * <p>
+     * Benchmark results show significant improvements:
+     * <ul>
+     * <li>Map size 5: 37% faster (90.9ns → 57.5ns)</li>
+     * <li>Map size 75: 47% faster (1248ns → 667ns)</li>
+     * <li>Map size 1000: 39% faster (18625ns → 11452ns)</li>
+     * </ul>
+     * </p>
+     *
+     * @return a mutable copy of the current thread context map, never {@code null}
```
```suggestion
     * {@return a mutable copy of the current thread context map}
```

vy commented on src/changelog/.2.x.x/3935_optimize_DefaultThreadContextMap_getCopy.xml:
```diff
@@ -0,0 +1,22 @@
+        
+        * 37% faster for maps with 5 elements (90.9ns → 57.5ns)
+        * 47% faster for maps with 75 elements (1248ns → 667ns) 
+        * 39% faster for maps with 1000 elements (18625ns → 11452ns)
+        
+        The change maintains identical functional behavior while eliminating megamorphic virtual calls that prevent JIT optimization.
+        This optimization particularly benefits applications with heavy ThreadContext usage such as web applications and microservices.
+    </description>
```
```suggestion
    <issue id="3935" link="https://github.com/apache/logging-log4j2/issues/3935"/>
    <issue id="3939" link="https://github.com/apache/logging-log4j2/pull/3939"/>
    <description format="asciidoc">
        Optimize `DefaultThreadContextMap.getCopy()` performance by avoiding megamorphic calls in `HashMap` constructor
    </description>
```

vy commented on log4j-api-test/src/test/java/org/apache/logging/log4j/spi/DefaultThreadContextMapTest.java:
```diff
@@ -130,4 +132,144 @@ void threadLocalNotInheritableByDefault() {
         threadLocalInheritableIfConfigured(createInheritableThreadContextMap());
     }
+
+    /**
+     * Test getCopy() with empty map
+     */
```
```suggestion
```

vy commented on log4j-api-test/src/test/java/org/apache/logging/log4j/spi/DefaultThreadContextMapTest.java:
```diff
@@ -130,4 +132,144 @@ void threadLocalNotInheritableByDefault() {
+        assertTrue(contextMap.isEmpty());
+    }
+
+    /**
+     * Test getCopy() with single-element map
+     */
```
```suggestion
```

vy commented on log4j-api-test/src/test/java/org/apache/logging/log4j/spi/DefaultThreadContextMapTest.java:
```diff
@@ -130,4 +132,144 @@ void threadLocalNotInheritableByDefault() {
+        assertThat(copy).isInstanceOf(HashMap.class);
+        assertEquals(5, copy.size());
+
+        for (Map.Entry<String, String> entry : testData.entrySet()) {
+            assertTrue(copy.containsKey(entry.getKey()));
+            assertEquals(entry.getValue(), copy.get(entry.getKey()));
+        }
+
```
Doesn't this duplicate `assertEquals(testData, copy)`?

```suggestion
```

anuragdy replied:
Good catch - you're right. The individual entry checks are redundant since `assertEquals(testData, copy)` already verifies the maps are equal. I'll remove the loop and keep the single assertion.

vy commented on log4j-api-test/src/test/java/org/apache/logging/log4j/spi/DefaultThreadContextMapTest.java:
```diff
@@ -130,4 +132,144 @@ void threadLocalNotInheritableByDefault() {
+        assertEquals(5, contextMap.size());
+    }
+
+    /**
+     * Test getCopy() returns proper HashMap type
+     */
```
```suggestion
```

vy commented on log4j-api-test/src/test/java/org/apache/logging/log4j/spi/DefaultThreadContextMapTest.java:
```diff
@@ -130,4 +132,144 @@ void threadLocalNotInheritableByDefault() {
+
+    /**
+     * Test getCopy() independence from original map
+     */
```
```suggestion
```

vy reviewed (changes requested):
@jengebr, @anuragdy, thanks so much for the report and the fix. I see that you provided a benchmark in #3935. Curious: Have you checked the existing `log4j-perf-test` benchmarks on thread context map? Are you able to reproduce the improvement figures using one of the benchmarks there? If so, which one(s)? If not, would you mind updating existing MDC benchmarks accordingly, please?

vy commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/DefaultThreadContextMap.java:
```diff
@@ -147,13 +147,38 @@ public <V> V getValue(final String key) {
+        final Map<String, String> map = getMap(state);
+
+        // Handle empty map case efficiently - constructor is faster for empty maps
+        // (manual iteration shows 10-23% regression for empty maps)
```
```suggestion
```

vy replied:
```suggestion
```

vy commented:
@anuragdy, build is failing, would you mind fixing it, please? See [the Log4j Development Guide](https://github.com/apache/logging-log4j2/blob/2.x/BUILDING.adoc#development) for details.

vy commented:
@anuragdy, please sign your commits.

anuragdy committed:
Merge branch 'apache:2.x' into 2.x

anuragdy committed:
Spotbug updates

anuragdy commented:
> @anuragdy, build is failing, would you mind fixing it, please? See [the Log4j Development Guide](https://github.com/apache/logging-log4j2/blob/2.x/BUILDING.adoc#development) for details.

Thanks for the guidelines, updated. The build workflow is succeeding on the fork now.
https://github.com/anuragdy/logging-log4j2/actions/runs/18404707294/job/52441832298

> @anuragdy, please sign your commits.

Updated the commits.

vy committed:
Remove redundant performance figure
